react/lesson2.md

```diff
@@ -1,82 +1,168 @@
-componentDidMount() {
-    this.callAPI();
+import React from 'react' // don't need to import Component
+
+function Profile(props) {
+  return (
+    <div>
+      Name: {props.name}
+      Age: {props.age}
+    </div>
+  )
+}
+
+export default Profile
+```
+
+### State
+
+To manage values that can change in your application - you store those values in a component's state. We **must** use class components if we want to store state, and to update the state we need to use `this.setState`, passing it the new object to put in our state. If we use this method to update our state - react will magically handle updating the html/output for us.
+
+[Click here](https://csb-vmxpvp07k0-xxlmpfwqxt.now.sh/) to view the next example in a browser. Try it out and understand what it's doing before we take a look at the code for it.
+
+Once you've done that - take a look at the implementation:
+
+```jsx
+import React, { Component } from 'react'
+
+class Counter extends Component {
+
+  // To initialise our state, we create a constructor and
+  // directly set the state.
+  // This is the **only** place we should ever write "this.state ="
+  constructor() {
+    super()
+    this.state = {
+      counter: 0
+    }
+  }
+
+  // this function will increase the number we have
+  // in our counter by 1 each time it is called
+  incrementCounter() {
+    // we're calling a React-provided method
+    // here to update our state
+    this.setState({
+      counter: this.state.counter + 1
+    })
+  }
+
+  // this function does the same as the above except it will
+  // reduce the number we have in the state by 1
+  decrementCounter() {
+    this.setState({
+      counter: this.state.counter - 1
+    })
+  }
+
+  render() {
+    // here we render out the current counter with
+    // two buttons, with onClicks that reference
+    // the functions we created above
+    return (
+      <div>
+        <br/>
+        <h1>Counter:{this.state.counter}</h1>
+        <button onClick={this.incrementCounter}>+</button>
+        <button onClick={this.decrementCounter}>-</button>
+      </div>
+    )
+  }
 }
 ```
 
-Once the state is updated with the data, your component will re-render.
+That's almost it for the recap, before we jump into the next part - have a go at trying to extend this counter example. Here's a few ideas that you could try:
 
-# Homework
-* [Lesson 2 Homework](Homework/Homework2.md)
+1. Change the colour of the counter, based on whether or not it's an even number ([demo here](https://csb-jopk2757y-mkxkadmfba.now.sh/))
+2. Add another button that multiplies the number by 2 ([demo here](https://csb-24pk29x2pn-zpnhuedswr.now.sh/))
+3. Add a slider that modifies that multiplier ([demo here](https://csb-l4kk6zn4kz-ypkmbtwnxy.now.sh/))
+  * So moving the slider to the left makes the multiplier 2, but moving it further to the right will increase the multiplier to 2,3,4 etc.
+4. Make the number of counters dynamic ([demo here](https://csb-nr8nvm9k70-fdvdaqcbst.now.sh/))
```

Comment:
Are those URLs permanent?

Reply:
Yeh - in theory anyway. They could go after ~6 months if they don't receive any traffic, it depends on how http://zeit.co/now handles that sort of thing.

Reply:
I'm getting a build error on https://csb-24pk29x2pn-zpnhuedswr.now.sh